iceberg_rust_ffi/src/catalog.rs

```diff
@@ -15,25 +17,138 @@ use object_store_ffi::{
 use crate::util::{parse_c_string, parse_properties, parse_string_array};
 use crate::PropertyEntry;
 
+/// Callback function type for custom token authentication from FFI
+///
+/// The callback receives:
+/// - user_data: opaque pointer to user context (e.g., Julia closure)
+/// - token_ptr: output pointer where the token string should be written
+///
+/// Returns:
+/// - 0 for success (token_ptr must point to a CString allocated with CString::into_raw)
+/// - non-zero for error
+pub type CustomAuthenticatorCallback =
+    extern "C" fn(user_data: *mut c_void, token_ptr: *mut *mut c_char) -> i32;
+
+/// Rust implementation of CustomAuthenticator that calls a C callback with user_data
+#[derive(Debug, Clone)]
+struct FFITokenAuthenticator {
+    callback: CustomAuthenticatorCallback,
+    user_data: *mut c_void,
+}
+
+// SAFETY: We trust that the Julia callback is thread-safe.
+// The user_data pointer is opaque and its thread-safety is the caller's responsibility.
+unsafe impl Send for FFITokenAuthenticator {}
+unsafe impl Sync for FFITokenAuthenticator {}
+
+#[async_trait]
+impl CustomAuthenticator for FFITokenAuthenticator {
+    async fn get_token(&self) -> iceberg::Result<String> {
+        let mut token_ptr: *mut c_char = std::ptr::null_mut();
+
+        let result = (self.callback)(self.user_data, &mut token_ptr);
```

Comment:
What happens if there's an exception in callback? I assume it's unsafe behaviour, are we guarded against it?